### PR TITLE
Merge HTTP and HTTPS traffic to the single port

### PR DESCRIPTION
**Pull request by soffokl**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -19,21 +19,17 @@ package main
...
 	"net/url"
 	"strings"
 
-	"github.com/inconshreveable/go-vhost"
 	"github.com/mysteriumnetwork/openvpn-forwarder/api"
 	"github.com/mysteriumnetwork/openvpn-forwarder/proxy"
 	netproxy "golang.org/x/net/proxy"
 )
 
+var proxyAddr = flag.String("proxy.bind", ":8080", "Proxy address for incoming connections")
```
Comment (Waldz) [inline main.go]: Port 8443, because it's HTTPS proxy - we connect to final endpoint) without proxying request/response itself

File: main.go [hunk 2]
```diff
@@ -103,63 +94,9 @@ func main() {
...
-		if err := proxyConnection.ConnectTo(conn, remoteHost, sm.Hash(clientHost)); err != nil {
-			log.Printf("Error establishing CONNECT tunnel to %s: %v", remoteHost, err)
-			return
-		}
-	}
-
-	go io.Copy(conn, tlsConn)
-	io.Copy(tlsConn, conn)
+	proxyServer := proxy.NewServer(*proxyAddr, dialer, sm, domainTracer)
+	log.Print("Serving proxy requests on ", *proxyAddr)
```
Comment (Waldz) [inline main.go]: "Serving HTTPS proxy on :8443"

File: proxy/server.go [hunk 3]
```diff
@@ -18,74 +18,135 @@
...
+	go s.handler(httpL, s.serveHTTP)
+	go s.handler(httpsL, s.serveTLS)
+
+	m.Serve()
+}
 
-		resp, err := http.ReadResponse(bufio.NewReader(conn), req)
+func (s *proxyServer) handler(l net.Listener, f func(c net.Conn)) {
+	for {
+		c, err := l.Accept()
```
Comment (Waldz) [inline proxy/server.go]: Some resource clean e.g. `defer c.Close()` needed.
Would be nice next to place where resources are allocated.

File: proxy/server.go [hunk 4]
```diff
@@ -18,74 +18,135 @@
...
+	sc := httputil.NewServerConn(c, nil)
+	req, err := sc.Read()
+	if err != nil {
+		log.Printf("Failed to read HTTP request: %v", err)
+		return
+	}
 
-		return req, resp
-	})
+	remoteHost := net.JoinHostPort(req.Host, "80")
```
Comment (Waldz) [inline proxy/server.go]: Is port somethere in request? Would be nice to find a way not to hardcode this port

File: proxy/server_test.go [hunk 5]
```diff
@@ -1,127 +0,0 @@
...
-	"net/url"
-	"sync"
-	"testing"
-
-	"net/http/httptest"
-
-	"github.com/stretchr/testify/assert"
-)
-
-func Test_Server_ServeHTTP(t *testing.T) {
```
Comment (Waldz) [inline proxy/server_test.go]: Would be nice put in place simple test for common case.
How hard would is that?

File: proxy/server.go [hunk 6]
```diff
@@ -18,74 +18,135 @@
...
-		req.URL.Host = req.Host
-		server.ServeHTTP(w, req)
-	})
-
-	server.ConnectDial = upstreamDialer.Dial
-	server.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
-		return goproxy.OkConnect, host
-	})
-
-	server.OnRequest().DoFunc(func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
```
Comment (Waldz) [inline proxy/server.go]: Goproxy was additionally serving 3rd endpoint - explicitly accepting CONNECT queries, which allowed to test proxying logic wo configured tunnel redirect with iptables.

It's documented here https://github.com/mysteriumnetwork/openvpn-forwarder

```
3. Check if forwarder is redirecting requests to upstream HTTPS proxy
FORWARDER_IP=`docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' forwarder`
curl --proxy $FORWARDER_IP:8080 "http://ipinfo.io"
```

File: main.go (hunk 1 above)
Comment (soffokl) [reply main.go]: Changed.

File: main.go (hunk 2 above)
Comment (soffokl) [reply main.go]: Changed.

File: proxy/server.go (hunk 4 above)
Comment (soffokl) [reply proxy/server.go]: Changed it to get port from an HTTP request.

File: proxy/server.go (hunk 3 above)
Comment (soffokl) [reply proxy/server.go]: Added closing it close to the `Accept`.

File: proxy/server.go [hunk 7]
```diff
@@ -18,74 +18,135 @@
...
+		return
+	}
+	defer tlsConn.Close()
+
+	if tlsConn.Host() == "" {
+		log.Printf("Cannot support non-SNI enabled TLS sessions")
+		return
+	}
+
+	remoteHost := net.JoinHostPort(tlsConn.Host(), "443")
```
[the remark this answers is not in the document]
Comment (soffokl) [reply proxy/server.go]: There is no port for the SNI request. Added explicit mapping for such cases.

File: proxy/server_test.go (hunk 5 above)
Comment (soffokl) [reply proxy/server_test.go]: Added a test for basic requests.

File: proxy/server.go (hunk 6 above)
Comment (soffokl) [reply proxy/server.go]: Added support for direct requests too.

File: proxy/server.go [hunk 8]
```diff
@@ -18,74 +18,135 @@
...
+func (s *proxyServer) serveTLS(c net.Conn) {
+	tlsConn, err := vhost.TLS(c)
+	if err != nil {
+		log.Printf("Error accepting new connection - %v", err)
+		return
+	}
+	defer tlsConn.Close()
+
+	if tlsConn.Host() == "" {
+		log.Printf("Cannot support non-SNI enabled TLS sessions")
```
[the remark this answers is not in the document]
Comment (soffokl) [reply proxy/server.go]: Created issue for it https://github.com/mysteriumnetwork/openvpn-forwarder/issues/7

File: proxy/server.go [hunk 9]
```diff
@@ -18,74 +18,168 @@
...
+		sm:      mapper,
+		dt:      dt,
+		portMap: portMap,
+	}
+}
+
+func (s *proxyServer) ListenAndServe(addr string) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("Error listening for https connections - %v", err)
```
Comment (Waldz) [inline proxy/server.go]: It's same as panicing in the middle of code, let's bubble errors

File: main.go [hunk 10]
```diff
@@ -103,63 +100,37 @@ func main() {
...
-			return
-		}
-		if err := proxyConnection.ConnectTo(conn, remoteHost, sm.Hash(clientHost)); err != nil {
-			log.Printf("Error establishing CONNECT tunnel to %s: %v", remoteHost, err)
-			return
+	portsMap := map[string]string{port: "443"}
+	for _, p := range ports {
+		portMap := strings.Split(p, ":")
+		if len(portMap) != 2 {
+			log.Fatalf("Failed to parse port mapping: %s", p)
```
[the remark this answers is not in the document]
Comment (soffokl) [reply main.go]: Done.

File: main.go [hunk 11]
```diff
@@ -19,26 +19,28 @@ package main
...
 var proxyAPIAddr = flag.String("proxy.api-bind", ":8000", "HTTP proxy API address")
-var proxyHTTPAddr = flag.String("proxy.http-bind", ":8080", "HTTP proxy address for incoming connections")
-var proxyHTTPSAddr = flag.String("proxy.https-bind", ":8443", "HTTPS proxy address for incoming connections")
 var proxyUpstreamURL = flag.String(
 	"proxy.upstream-url",
 	"",
 	`Upstream HTTPS proxy where to forward traffic (e.g. "http://superproxy.com:8080")`,
 )
+var proxyMapPort = FlagArray(
+	"proxy.map.port",
```
[the remark this answers is not in the document]
Comment (soffokl) [reply main.go]: Done.

File: proxy/server.go (hunk 9 above)
Comment (soffokl) [reply proxy/server.go]: Done.

**Pull request merged**